src/garage/tf/q_functions/continuous_mlp_q_function_with_model.py

```diff
@@ -0,0 +1,167 @@
+                 output_nonlinearity=None,
+                 output_w_init=tf.glorot_uniform_initializer(),
+                 output_b_init=tf.zeros_initializer(),
+                 input_include_goal=False,
+                 layer_normalization=False):
+        super().__init__(name)
+
+        self._env_spec = env_spec
+        self._hidden_sizes = hidden_sizes
+        self._action_merge_layer = action_merge_layer
+        self._hidden_nonlinearity = hidden_nonlinearity
+        self._hidden_w_init = hidden_w_init
+        self._hidden_b_init = hidden_b_init
+        self._output_nonlinearity = output_nonlinearity
+        self._output_w_init = output_w_init
+        self._output_b_init = output_b_init
+        self._input_include_goal = input_include_goal
+        self._layer_normalization = layer_normalization
+
+        if self._input_include_goal:
+            self._obs_dim = env_spec.observation_space.flat_dim_with_keys(
+                ['observation', 'desired_goal'])
+        else:
+            self._obs_dim = env_spec.observation_space.flat_dim
+        self._action_dim = env_spec.action_space.flat_dim
+
+        self.model = MLPMergeModel(
+            output_dim=1,
+            hidden_sizes=hidden_sizes,
+            concat_layer=self._action_merge_layer,
+            hidden_nonlinearity=hidden_nonlinearity,
+            hidden_w_init=hidden_w_init,
+            hidden_b_init=hidden_b_init,
+            output_nonlinearity=output_nonlinearity,
+            output_w_init=output_w_init,
+            output_b_init=output_b_init,
+            layer_normalization=layer_normalization)
+
+        self._initialize()
+
+    def _initialize(self):
+        obs_ph = tf.placeholder(tf.float32, (None, self._obs_dim), name='obs')
+        action_ph = tf.placeholder(
+            tf.float32, (None, self._action_dim), name='act')
+
+        with tf.variable_scope(self.name) as vs:
+            self._variable_scope = vs
+            self.model.build(obs_ph, action_ph)
+
+        self._f_qval = tf.get_default_session().make_callable(
+            self.model.networks['default'].outputs,
+            feed_list=[obs_ph, action_ph])
+
+    def get_qval(self, observation, action):
+        """Q Value of the network."""
+        return self._f_qval(observation, action)
+
+    @property
+    def inputs(self):
+        """Return the input tensor."""
+        return self.model.networks['default'].inputs
+
+    @overrides
+    def get_qval_sym(self, state_input, action_input, name):
+        """
+        Symbolic graph for q-network.
+
+        Args:
+            state_input (tf.Tensor): The state input tf.Tensor to the network.
+            name (str): Network variable scope.
+
+        Return:
+            The tf.Tensor output of Discrete MLP QFunction.
+        """
+        with tf.variable_scope(self._variable_scope):
+            return self.model.build(state_input, action_input, name=name)
+
+    def clone(self, name):
```

Comment:
This is definitely a code-smell. Basically, this function is doing the same work that setting a name in the `Model.build` method does, but with a different API.

Unfortunately, how to fix this in a good way is out of scope for this PR, so let's leave it like this for now.

Reply:
this is a well-known flaw caused by having the proxy through the Policy interface (but Policy is not a model).

this can be addressed by inherting Policies directly from Model, which is what we plan to do once we have removed layers. if there is no issue already, please file one.